Object filter types

PascalSenn opened the pull request:
Current implementation is limited to schema generation. 

I didn't know how to lazy initialize a type 
Check out TODO comments in ObjectFilterFieldDescriptor.cs

Addresses: #921

PascalSenn commented on src/Core/Types.Filters/Object/ObjectFilterFieldDescriptor.cs:
```diff
@@ -0,0 +1,87 @@
+            var desc = FilterInputTypeDescriptor<TObject>.New(Context, typeof(TObject));
+            descriptor.Invoke(desc);
+
+            // TODO: I don't know how to initialize this type 
```
@michaelstaib  Here

michaelstaib replied:
What is the invoke doing?

michaelstaib replied:
ah .... I see .... to early in the  morning ;)

PascalSenn commented on src/Core/Types.Filters/Object/ObjectFilterFieldDescriptor.cs:
```diff
@@ -0,0 +1,87 @@
+        public IObjectFilterFieldDescriptor<TObject> AllowObject<TFilter>() where TFilter : IFilterInputTypeDescriptor<TObject>
+        {
+
+            // TODO: Do wee need to pass the generic TFilter to create Option?
```
@michaelstaib And here

michaelstaib commented:
OK, id did some more work to make extending types easier....

michaelstaib commented:
```csharp
        public IObjectFilterFieldDescriptor<TObject> AllowObject(
            Action<IFilterInputTypeDescriptor<TObject>> descriptor)
        {
            var type = new FilterInputType<TObject>(descriptor);
            var typeReference = new SchemaTypeReference(type);
            .... 
            return this;
        }
```

You still have now to rewrite the following part:

```csharp
            ObjectFilterOperationDescriptor<TObject> field =
                CreateOperation(FilterOperationKind.Equals);
            Filters.Add(field);
```

Do we need to add a new operation kind for that?

The `typeReference` is the return type of the operation field so this would need to become the `field.Type(typeReference)`.

PascalSenn commented on src/Core/Types.Filters/FilterInputTypeDescriptor.cs:
```diff
@@ -116,6 +116,14 @@ public class FilterInputTypeDescriptor<T>
                 return true;
             }
 
+            if (property.PropertyType.IsClass)
```
Is this gate safe?

PascalSenn commented on src/Core/Types.Filters/FilterInputTypeDescriptor.cs:
```diff
@@ -183,6 +191,22 @@ public class FilterInputTypeDescriptor<T>
         }
 
+        public IObjectFilterFieldDescriptor<TObject> Filter<TObject>(
+            Expression<Func<T, TObject>> propertyOrMethod) where TObject : class
```
I wonder if there is a better generic constraint than this...

michaelstaib replied:
And also what about interfaces or structs that represent a complex type.

michaelstaib replied:
Maybe the name is wrong.... is it really a filter? I mean in the case of an object?

michaelstaib replied:
And what if an object is comparable :D

PascalSenn commented on src/Core/Types.Filters/Object/ObjectFilterFieldDescriptor.cs:
```diff
@@ -0,0 +1,76 @@
+        public ObjectFilterFieldDescriptor(
+            IDescriptorContext context,
+            PropertyInfo property,
+            Type type)
```
Should i remove type and just read the type over the property?

michaelstaib replied:
I will have to check.

PascalSenn commented on src/Core/Types/Types/Descriptors/ArgumentDescriptorBase~1.cs:
```diff
@@ -63,6 +63,17 @@ public void Type<TInputType>(TInputType inputType)
             Definition.Type = new SchemaTypeReference(inputType);
         }
 
+        public void Type(ITypeReference typeReference)
```
Does this makes sense?

michaelstaib replied:
Not on public stuff .... the ITypeReference is only for internal purposes ... they basically also are placeholders for the actual type while the schema builder is creating the schema.

michaelstaib replied:
why do you need it there anyway?

PascalSenn commented on src/Core/Types.Filters/Object/ObjectFilterFieldDescriptor.cs:
```diff
@@ -0,0 +1,76 @@
+                Context,
+                this,
+                CreateFieldName(operationKind),
+                new ClrTypeReference(typeof(FilterInputType<>).MakeGenericType(type), Definition.Type.Context, true, true),
```
hm......

PascalSenn commented on src/Core/Types.Filters/Object/ObjectFilterFieldDescriptor~1.cs:
```diff
@@ -0,0 +1,89 @@
+            return field;
+        }
+
+        public new IObjectFilterOperationDescriptor<TObject> AllowObject()
```
I introduced this API for implicit bindings

PascalSenn commented:
@michaelstaib  Thanks, helped a lot. I could resolve the schema and added a few tests. 
I do have some open questions. I added reviews for them. Basically i highlighted the parts I'm not sure if I've done everything right 👍 

I guess after this review we can start to work on the visitor

michaelstaib commented:
I am focusing at the moment to finish 10x tasks... after I have finished those I will focus in this one with you.

michaelstaib commented:
Is there something blocking or can we get this one going again?

PascalSenn commented:
I had a look in it a few days ago. 
I was trying to figure out the best way for the visitor. There is still that comment you left there for object types. I guess you already knoe exactly how to do it?

Type generation seems to work. You may want to have a closer look at the PR. There are a few things that are not very nice.

PascalSenn commented:
@michaelstaib  I'm kind of stuck on the visitor.  Once this is done i could start on the tests

michaelstaib commented:
I will help you tomorrow on that one. Sorry for the wait.

michaelstaib commented:
@PascalSenn I have to fix  the circle/ci ... the container still uses .net core 3 preview .... 

I will fix it and hit rebuild.

michaelstaib commented:
OK, pushed the build fix.

michaelstaib commented:
Circle CI is green now ....

michaelstaib commented:
I wait for the sonar results then I start the review.

michaelstaib commented:
I have fixed some sonar issues and I think we can merge it once it is green. We will test it internally and see if it has what it takes.

If we find issues we can still iterate on it.
